src/platform/packages/shared/kbn-esql-utils/src/utils/cascaded_documents_helpers.ts

```diff
@@ -0,0 +1,489 @@
+        }
+
+        return Builder.expression.func.call('match', [
+          Builder.identifier({ name: (arg as ESQLColumn).text }),
+          Builder.expression.literal.string(extractCategorizeTokens(matchValue).join(' ')),
+          Builder.expression.map({
+            entries: [
+              Builder.expression.entry(
+                'auto_generate_synonyms_phrase_query',
+                Builder.expression.literal.boolean(false)
+              ),
+              Builder.expression.entry('fuzziness', Builder.expression.literal.integer(0)),
+              Builder.expression.entry('operator', Builder.expression.literal.string('AND')),
+            ],
+          }),
+        ]);
+      })
+      .filter(Boolean) as ESQLAstItem[],
+  });
+
+  mutate.generic.commands.append(cascadeOperationQuery.ast, categorizeWhereCommand);
+
+  return {
+    esql: BasicPrettyPrinter.print(cascadeOperationQuery.ast),
+  };
+}
+
+/**
+ * Modifies the provided ESQL query to only include the specified columns in the stats by option.
+ */
+export function mutateQueryStatsGrouping(query: AggregateQuery, pick: string[]): AggregateQuery {
+  const EditorESQLQuery = EsqlQuery.fromSrc(query.esql);
+
+  const dataSourceCommand = mutate.generic.commands.find(
+    EditorESQLQuery.ast,
+    (cmd) => cmd.name === 'from'
+  ) as ESQLCommand<'from'> | undefined;
+
+  if (!dataSourceCommand) {
+    throw new Error('Query does not have a data source');
+  }
+
+  const statsCommands = Array.from(mutate.commands.stats.list(EditorESQLQuery.ast));
+
+  if (statsCommands.length === 0) {
+    throw new Error(`Query does not include a "stats" command`);
+  }
+
+  let statsCommandToOperateOn: StatsCommand | null = null;
+  let statsCommandToOperateOnGrouping: StatsCommandSummary['grouping'] | null = null;
+
+  // accounting for the possibility of multiple stats commands in the query,
+  // we always want to operate on the last stats command that has valid grouping options
+  for (let i = statsCommands.length - 1; i >= 0; i--) {
+    ({ grouping: statsCommandToOperateOnGrouping } = mutate.commands.stats.summarizeCommand(
+      EditorESQLQuery,
+      statsCommands[i]
+    ));
+
+    if (statsCommandToOperateOnGrouping && Object.keys(statsCommandToOperateOnGrouping).length) {
+      statsCommandToOperateOn = statsCommands[i] as StatsCommand;
+      break;
+    }
+  }
+
+  if (!statsCommandToOperateOn) {
+    throw new Error(`No valid "stats" command was found in the query`);
+  }
+
+  const isValidPick = pick.every(
+    (col) =>
+      Object.keys(statsCommandToOperateOnGrouping!).includes(col) ||
+      Object.keys(statsCommandToOperateOnGrouping!).includes(`\`${col}\``)
+  );
+
+  if (!isValidPick) {
+    // nothing to do, return query as is
+    return {
+      esql: BasicPrettyPrinter.print(EditorESQLQuery.ast),
+    };
+  }
+
+  // Create a modified stats command with only the specified column as args for the "by" option
+  const modifiedStatsCommand = Builder.command({
```

Comment:
Hmm I wonder if we can do it smarter (and with less code) with the composer api. @vadimkibana can you take a look?